python/gps/utility/visualization.py

```diff
@@ -0,0 +1,141 @@
+from gps.proto.gps_pb2 import END_EFFECTOR_POINTS, RGB_IMAGE
+import numpy as np
+import matplotlib.pyplot as plt
+
+TRIALS = 20
+THRESH = {'reacher': 0.05, 'pointmass': 0.1}
+
+def compare_samples_curve(gps, N, agent_config, weight_varying=False, experiment='reacher'):
+    """
+    Compare samples between IOC and demo policies and visualize them in a plot.
+    Args:
+        gps: GPS object.
+        N: number of samples taken from the policy for comparison
+        config: Configuration of the agent to sample.
+        weight_varying: whether the experiment is weight-varying or not.
+        experiment: whether the experiment is reacher or pointmass.
+    """
+    pol_iter = gps._hyperparams['algorithm']['iterations'] - 1
+    algorithm_ioc = gps.data_logger.unpickle(gps._data_files_dir + 'algorithm_itr_%02d' % pol_iter + '.pkl')
+    algorithm_sup = gps.data_logger.unpickle(gps._hyperparams['common']['supervised_exp_dir'] + 'data_files/algorithm_itr_%02d' % pol_iter + '.pkl')
+    algorithm_demo = gps.data_logger.unpickle(gps._hyperparams['common']['demo_exp_dir'] + 'data_files/algorithm_itr_09.pkl') # Assuming not using 4 policies
+    algorithm_oracle = gps.data_logger.unpickle(gps._hyperparams['common']['demo_exp_dir'] + 'data_files_oracle/algorithm_itr_09.pkl')
+    if not weight_varying:
+        pos_body_offset = gps._hyperparams['agent']['pos_body_offset']
+    M = agent_config['conditions']
+    if experiment == 'reacher' and not weight_varying: #reset body offsets
+        np.random.seed(101)
+        for m in range(M):
+            self.agent.reset_initial_body_offset(m)
+
+    pol_ioc = algorithm_ioc.policy_opt.policy
+    pol_sup = algorithm_sup.policy_opt.policy
+    pol_demo = algorithm_demo.policy_opt.policy
+    pol_oracle = algorithm_oracle.policy_opt.policy
+    policies = [pol_ioc, pol_sup, pol_demo, pol_oracle]
+    successes = {i: np.zeros((TRIALS, M)) for i in xrange(len(policies))}
+    agent = agent_config['type'](agent_config)
+    if not weight_varying:
+        ioc_conditions = agent_config['pos_body_offset']
+    else:
+        ioc_conditions = [np.log10(agent_config['density_range'][i])-4.0 \
+                            for i in xrange(M)]
+    pos_body_offset = gps.agent._hyperparams['pos_body_offset'][i]
+    target_position = np.array([.1,-.1,.01])+pos_body_offset
+
+    import random
+
+    for seed in xrange(TRIALS):
+        random.seed(seed)
+        np.random.seed(seed)
+        for i in xrange(M):
+            # Gather demos.
+            for j in xrange(N):
+                for k in xrange(len(policies)):
+                    sample = agent.sample(
+                        policies[k], i,
+                        verbose=(i < gps._hyperparams['verbose_trials']), noisy=True
```

Comment:
put noisy on next line to be within line length limit (and since it is already broken up)